rhino/src/main/java/org/mozilla/javascript/SlotMapContainer.java

```diff
@@ -55,17 +58,112 @@ public Slot query(Object key, int index) {
-                SlotMapOwner owner, Object key, int index, SlotComputer<S> compute) {
-            var map = new EmbeddedSlotMap();
-            owner.setMap(map);
-            return map.compute(owner, key, index, compute);
+                SlotMapOwner owner, Object key, int index, SlotComputer<S> c) {
+            var newSlot = c.compute(key, index, null);
+            if (newSlot != null) {
+                var map = new SingleEntrySlotMap(newSlot);
+                owner.setMap(map);
+            }
+            return newSlot;
+        }
+    }
+
+    private static final class Iter implements Iterator<Slot> {
+        private Slot next;
+
+        Iter(Slot slot) {
+            next = slot;
+        }
+
+        @Override
+        public boolean hasNext() {
+            return next != null;
+        }
+
+        @Override
+        public Slot next() {
+            Slot ret = next;
+            if (ret == null) {
+                throw new NoSuchElementException();
+            }
+            next = next.orderedNext;
+            return ret;
+        }
+    }
+
+    static final class SingleEntrySlotMap implements SlotMap {
+
+        SingleEntrySlotMap(Slot slot) {
+            assert (slot != null);
+            this.slot = slot;
+        }
+
+        private final Slot slot;
+
+        @Override
+        public Iterator<Slot> iterator() {
+            return new Iter(slot);
+        }
+
+        @Override
+        public int size() {
+            return 1;
+        }
+
+        @Override
+        public boolean isEmpty() {
+            return false;
+        }
+
+        @Override
+        public Slot modify(SlotMapOwner owner, Object key, int index, int attributes) {
+            final int indexOrHash = (key != null ? key.hashCode() : index);
+
+            if (indexOrHash == slot.indexOrHash && Objects.equals(slot.name, key)) {
+                return slot;
+            }
+            Slot newSlot = new Slot(key, index, attributes);
+            add(owner, newSlot);
+            return newSlot;
+        }
+
+        @Override
+        public Slot query(Object key, int index) {
+            final int indexOrHash = (key != null ? key.hashCode() : index);
+
+            if (indexOrHash == slot.indexOrHash && Objects.equals(slot.name, key)) {
+                return slot;
+            }
+            return null;
+        }
+
+        @Override
+        public void add(SlotMapOwner owner, Slot newSlot) {
+            if (owner == null) {
+                throw new IllegalStateException();
+            } else {
+                var newMap = new EmbeddedSlotMap();
+                owner.setMap(newMap);
+                newMap.add(owner, slot);
+                newMap.add(owner, newSlot);
+            }
+        }
+
+        @Override
+        public <S extends Slot> S compute(
+                SlotMapOwner owner, Object key, int index, SlotComputer<S> c) {
+            var newMap = new EmbeddedSlotMap();
```

Comment:
In theory, "compute" is also used to replace an existing slot, rather than add a new one. I don't know how much that happens in actual situations, but there are plenty of places in the codebase where it happens. With this implementation, in that case we'll upgrade to a "real" EmbeddedSlotMap, but then it still may have a size of one.

Would it complicate things a lot to support that use case? If not, should we try it? I would also be open to the idea that this would be a premature optimization.

Reply:
I chose to always promote there because I figured that a map which is mutated is likely to continue to be mutated, and it made it easier to think about the multithreaded cases.

Removing a slot also causes the map to be promoted as that maintains a strict type lattice of slot map types (we never demote a map). I tried to support that sort of thing initially but it made the code much more complex and was rarely triggered in my testing so I went for the simpler approach seen here.

Ultimately I think we should separate values and slot descriptors, and make the latter immutable. That would open up some good optimization opportunities in both the interpreter and compiler, and would lI think save quite a bit of memory.